Tests/NIOPosixTests/IPv4Header.swift

```diff
@@ -0,0 +1,277 @@
+        get {
+            versionAndIhl & 0b0000_1111
+        }
+        set {
+            precondition(newValue & 0b1111_0000 == 0)
+            versionAndIhl = newValue | (0b1111_0000 & versionAndIhl)
+            assert(newValue == internetHeaderLength)
+        }
+    }
+    private var dscpAndEcn: UInt8
+    var differentiatedServicesCodePoint: UInt8 {
+        get {
+            dscpAndEcn >> 2
+        }
+        set {
+            precondition(newValue & 0b0000_0011 == 0)
+            dscpAndEcn = newValue << 2 | (0b0000_0011 & dscpAndEcn)
+            assert(newValue == differentiatedServicesCodePoint)
+        }
+    }
+    var explicitCongestionNotification: UInt8 {
+        get {
+            dscpAndEcn & 0b0000_0011
+        }
+        set {
+            precondition(newValue & 0b0000_0011 == 0)
+            dscpAndEcn = newValue | (0b1111_1100 & dscpAndEcn)
+            assert(newValue == explicitCongestionNotification)
+        }
+    }
+    var totalLength: UInt16
+    var identification: UInt16
+    private var flagsAndFragmentOffset: UInt16
+    var flags: UInt8 {
+        get {
+            UInt8(flagsAndFragmentOffset >> 13)
+        }
+        set {
+            precondition(newValue & 0b0000_0111 == 0)
+            flagsAndFragmentOffset = UInt16(newValue) << 13 | (0b0001_1111_1111_1111 & flagsAndFragmentOffset)
+            assert(newValue == flags)
+        }
+    }
+    var fragmentOffset: UInt16 {
+        get {
+            flagsAndFragmentOffset & 0b0001_1111_1111_1111
+        }
+        set {
+            precondition(newValue & 0b1110_0000_0000_0000 == 0)
+            flagsAndFragmentOffset = newValue | (0b1110_0000_0000_0000 & flagsAndFragmentOffset)
+            assert(newValue == fragmentOffset)
+        }
+    }
+    var timeToLive: UInt8
+    var `protocol`: NIOIPProtocol
+    var headerChecksum: UInt16
+    var sourceIpAddress: IPv4Address
+    var destinationIpAddress: IPv4Address
+    
+    init?(buffer: inout ByteBuffer) {
+        #if canImport(Darwin)
+        guard
+            let versionAndIhl: UInt8 = buffer.readInteger(),
+            let dscpAndEcn: UInt8 = buffer.readInteger(),
+            // On BSD, the total length is in host byte order
+            let totalLength: UInt16 = buffer.readInteger(endianness: .host),
+            let identification: UInt16 = buffer.readInteger(),
+            // fragmentOffset is in host byte order as well but it is always zero in our tests
+            // and fragmentOffset is 13 bits in size so we can't just use readInteger(endianness: .host)
+            let flagsAndFragmentOffset: UInt16 = buffer.readInteger(),
+            let timeToLive: UInt8 = buffer.readInteger(),
+            let `protocol`: UInt8 = buffer.readInteger(),
+            let headerChecksum: UInt16 = buffer.readInteger(),
+            let sourceIpAddress: UInt32 = buffer.readInteger(),
+            let destinationIpAddress: UInt32 = buffer.readInteger()
+        else { return nil }
```

Comment:
this is incorrect because it forgets to reset the `buffer` if you half-parsed something.

Also this doesn't follow the usual pattern of `read*`/`write*` methods. And lastly, there's an implementation of this already:
- check `PCAP.swift` on this PR: https://github.com/apple/swift-nio-extras/pull/158 : https://github.com/apple/swift-nio-extras/pull/158/files#diff-8b0f17fa27d694721788c7adab20151f44ff3672becc16dd728bd530c0de44ba
- also here (this code is changed by the above PR): https://github.com/apple/swift-nio-extras/blob/main/Tests/NIOExtrasTests/WritePCAPHandlerTest.swift#L746

Reply:
Done in https://github.com/apple/swift-nio/pull/2320/commits/20ed3802642a475b760beb33d8a121430f0224a6